### ISSUE TITLE
Bug in merging populations from userscript and prototypecontext

### ISSUE THREAD

**Issue by stefjoosten** [AmpersandTarski/Ampersand#1035]
#### Version of ampersand that was used

I built RAP3 from Ampersand commit  AmpersandTarski/Ampersand@5010d66c5a3798249fecf8a886e90f7cf8852db0, which uses the development branch of Ampersand 4.0.0.

#### What I expected
I expected issue #1033 to be resolved and I expected RAP to appear on http://rap.localhost. In fact, I closed #1033 because it seems to be resolved.

#### What happened instead

After starting completely afresh (even flushing the browser cache) I got the following result (Screenshot):
<img width="1141" alt="Schermafdruk 2020-01-02 14 28 14" src="https://user-images.githubusercontent.com/275768/71669228-42e14980-2d6c-11ea-9bf4-b1ff06d8cdd7.png">

I looked at the logfile `/var/www/log/execengine-2020-01-02.log` in the `rap3`-container, finding the transaction that failed. It actually fixed three rules (all from `Navbar.adl` in the Ampersand compiler):
 - 7 violations in rule 'Add navItemRoles for interfaces'
 - 5 violations in rule 'Add navItemRoles for parent items'
 - 9 violations in rule 'Fix TOT for isPartOf[PF_NavMenuItem*PF_NavMenu]'
Then the exec engine claims that there are invariant rules that do not hold, and rolls back.
As a consequence, the navbar shows no menu items at all. Here is the relevant part of the log file:

```
[2020-01-02 14:52:45] TRANSACTION.INFO: Opening transaction: 883871315 [] []
[2020-01-02 14:52:45] TRANSACTION.DEBUG: Mark concept 'PF_NavMenu' as affected concept [] []
[2020-01-02 14:52:45] TRANSACTION.DEBUG: Add storage: MySQL database db - RAP3 [] []
[2020-01-02 14:52:45] TRANSACTION.DEBUG: Mark relation 'label[PF_NavMenuItem*PF_Label]' as affected relation [] []
[2020-01-02 14:52:45] TRANSACTION.DEBUG: Mark concept 'PF_Label' as affected concept [] []
[2020-01-02 14:52:45] TRANSACTION.DEBUG: Mark relation 'isVisible[PF_NavMenuItem*PF_NavMenuItem]' as affected relation [] []
[2020-01-02 14:52:45] TRANSACTION.DEBUG: Mark relation 'isPartOf[PF_NavMenuItem*PF_NavMenu]' as affected relation [] []
[2020-01-02 14:52:45] TRANSACTION.DEBUG: Mark concept 'PF_NavMenuItem' as affected concept [] []
[2020-01-02 14:52:45] TRANSACTION.DEBUG: Mark relation 'ifc[PF_NavMenuItem*PF_Interface]' as affected relation [] []
[2020-01-02 14:52:45] TRANSACTION.DEBUG: Mark relation 'seqNr[PF_NavMenuItem*PF_SeqNr]' as affected relation [] []
[2020-01-02 14:52:45] TRANSACTION.DEBUG: Mark concept 'PF_SeqNr' as affected concept [] []
[2020-01-02 14:52:45] TRANSACTION.DEBUG: Mark relation 'isSubItemOf[PF_NavMenuItem*PF_NavMenuItem]' as affected relation [] []
[2020-01-02 14:52:45] TRANSACTION.DEBUG: Mark concept 'PF_Interface' as affected concept [] []
[2020-01-02 14:52:45] EXECENGINE.INFO: ExecEngine started [] []
[2020-01-02 14:52:45] EXECENGINE.INFO: {+ Run #{1} (auto rerun: true) [] []
[2020-01-02 14:52:45] EXECENGINE.DEBUG: Select exec engine '{ExecEngine}' [] []
[2020-01-02 14:52:45] EXECENGINE.INFO: {++ ExecEngine fixing 7 violations for rule 'Add navItemRoles for interfaces' [] []
[2020-01-02 14:52:45] EXECENGINE.INFO: {+++ Fixing violation 1/7: ((Accounts[PF_NavMenuItem],AccountManager[Role])) [] []
[2020-01-02 14:52:45] EXECENGINE.INFO: InsPair(pf_navItemRoles,PF_NavMenuItem,Accounts,Role,AccountManager) [] []
[2020-01-02 14:52:45] TRANSACTION.DEBUG: Mark relation 'pf_navItemRoles[PF_NavMenuItem*Role]' as affected relation [] []
[2020-01-02 14:52:45] EXECENGINE.DEBUG: Added (Accounts[PF_NavMenuItem],AccountManager[Role])[pf_navItemRoles[PF_NavMenuItem*Role]] [] []
[2020-01-02 14:52:45] EXECENGINE.INFO: +++} [] []
[2020-01-02 14:52:45] EXECENGINE.INFO: {+++ Fixing violation 2/7: ((Login[PF_NavMenuItem],Anonymous[Role])) [] []
[2020-01-02 14:52:45] EXECENGINE.INFO: InsPair(pf_navItemRoles,PF_NavMenuItem,Login,Role,Anonymous) [] []
[2020-01-02 14:52:45] EXECENGINE.DEBUG: Added (Login[PF_NavMenuItem],Anonymous[Role])[pf_navItemRoles[PF_NavMenuItem*Role]] [] []
[2020-01-02 14:52:45] EXECENGINE.INFO: +++} [] []
[2020-01-02 14:52:45] EXECENGINE.INFO: {+++ Fixing violation 3/7: ((Logout[PF_NavMenuItem],User[Role])) [] []
[2020-01-02 14:52:45] EXECENGINE.INFO: InsPair(pf_navItemRoles,PF_NavMenuItem,Logout,Role,User) [] []
[2020-01-02 14:52:45] EXECENGINE.DEBUG: Added (Logout[PF_NavMenuItem],User[Role])[pf_navItemRoles[PF_NavMenuItem*Role]] [] []
[2020-01-02 14:52:45] EXECENGINE.INFO: +++} [] []
[2020-01-02 14:52:45] EXECENGINE.INFO: {+++ Fixing violation 4/7: ((MyAccount[PF_NavMenuItem],User[Role])) [] []
[2020-01-02 14:52:45] EXECENGINE.INFO: InsPair(pf_navItemRoles,PF_NavMenuItem,MyAccount,Role,User) [] []
[2020-01-02 14:52:45] EXECENGINE.DEBUG: Added (MyAccount[PF_NavMenuItem],User[Role])[pf_navItemRoles[PF_NavMenuItem*Role]] [] []
[2020-01-02 14:52:45] EXECENGINE.INFO: +++} [] []
[2020-01-02 14:52:45] EXECENGINE.INFO: {+++ Fixing violation 5/7: ((MyScripts[PF_NavMenuItem],Student[Role])) [] []
[2020-01-02 14:52:45] EXECENGINE.INFO: InsPair(pf_navItemRoles,PF_NavMenuItem,MyScripts,Role,Student) [] []
[2020-01-02 14:52:45] EXECENGINE.DEBUG: Added (MyScripts[PF_NavMenuItem],Student[Role])[pf_navItemRoles[PF_NavMenuItem*Role]] [] []
[2020-01-02 14:52:45] EXECENGINE.INFO: +++} [] []
[2020-01-02 14:52:45] EXECENGINE.INFO: {+++ Fixing violation 6/7: ((Roles[PF_NavMenuItem],AccountManager[Role])) [] []
[2020-01-02 14:52:45] EXECENGINE.INFO: InsPair(pf_navItemRoles,PF_NavMenuItem,Roles,Role,AccountManager) [] []
[2020-01-02 14:52:45] EXECENGINE.DEBUG: Added (Roles[PF_NavMenuItem],AccountManager[Role])[pf_navItemRoles[PF_NavMenuItem*Role]] [] []
[2020-01-02 14:52:45] EXECENGINE.INFO: +++} [] []
[2020-01-02 14:52:45] EXECENGINE.INFO: {+++ Fixing violation 7/7: ((StudentScripts[PF_NavMenuItem],Tutor[Role])) [] []
[2020-01-02 14:52:45] EXECENGINE.INFO: InsPair(pf_navItemRoles,PF_NavMenuItem,StudentScripts,Role,Tutor) [] []
[2020-01-02 14:52:45] EXECENGINE.DEBUG: Added (StudentScripts[PF_NavMenuItem],Tutor[Role])[pf_navItemRoles[PF_NavMenuItem*Role]] [] []
[2020-01-02 14:52:45] EXECENGINE.INFO: +++} [] []
[2020-01-02 14:52:45] EXECENGINE.INFO: ++} ExecEngine fixed 7 violations for rule 'Add navItemRoles for interfaces' [] []
[2020-01-02 14:52:45] EXECENGINE.INFO: {++ ExecEngine fixing 5 violations for rule 'Add navItemRoles for parent items' [] []
[2020-01-02 14:52:45] EXECENGINE.INFO: {+++ Fixing violation 1/5: ((MainMenu[PF_NavMenuItem],AccountManager[Role])) [] []
[2020-01-02 14:52:45] EXECENGINE.INFO: InsPair(pf_navItemRoles,PF_NavMenuItem,MainMenu,Role,AccountManager) [] []
[2020-01-02 14:52:45] EXECENGINE.DEBUG: Added (MainMenu[PF_NavMenuItem],AccountManager[Role])[pf_navItemRoles[PF_NavMenuItem*Role]] [] []
[2020-01-02 14:52:45] EXECENGINE.INFO: +++} [] []
[2020-01-02 14:52:45] EXECENGINE.INFO: {+++ Fixing violation 2/5: ((MainMenu[PF_NavMenuItem],Anonymous[Role])) [] []
[2020-01-02 14:52:45] EXECENGINE.INFO: InsPair(pf_navItemRoles,PF_NavMenuItem,MainMenu,Role,Anonymous) [] []
[2020-01-02 14:52:45] EXECENGINE.DEBUG: Added (MainMenu[PF_NavMenuItem],Anonymous[Role])[pf_navItemRoles[PF_NavMenuItem*Role]] [] []
[2020-01-02 14:52:45] EXECENGINE.INFO: +++} [] []
[2020-01-02 14:52:45] EXECENGINE.INFO: {+++ Fixing violation 3/5: ((MainMenu[PF_NavMenuItem],User[Role])) [] []
[2020-01-02 14:52:45] EXECENGINE.INFO: InsPair(pf_navItemRoles,PF_NavMenuItem,MainMenu,Role,User) [] []
[2020-01-02 14:52:45] EXECENGINE.DEBUG: Added (MainMenu[PF_NavMenuItem],User[Role])[pf_navItemRoles[PF_NavMenuItem*Role]] [] []
[2020-01-02 14:52:45] EXECENGINE.INFO: +++} [] []
[2020-01-02 14:52:45] EXECENGINE.INFO: {+++ Fixing violation 4/5: ((MainMenu[PF_NavMenuItem],Student[Role])) [] []
[2020-01-02 14:52:45] EXECENGINE.INFO: InsPair(pf_navItemRoles,PF_NavMenuItem,MainMenu,Role,Student) [] []
[2020-01-02 14:52:45] EXECENGINE.DEBUG: Added (MainMenu[PF_NavMenuItem],Student[Role])[pf_navItemRoles[PF_NavMenuItem*Role]] [] []
[2020-01-02 14:52:45] EXECENGINE.INFO: +++} [] []
[2020-01-02 14:52:45] EXECENGINE.INFO: {+++ Fixing violation 5/5: ((MainMenu[PF_NavMenuItem],Tutor[Role])) [] []
[2020-01-02 14:52:45] EXECENGINE.INFO: InsPair(pf_navItemRoles,PF_NavMenuItem,MainMenu,Role,Tutor) [] []
[2020-01-02 14:52:45] EXECENGINE.DEBUG: Added (MainMenu[PF_NavMenuItem],Tutor[Role])[pf_navItemRoles[PF_NavMenuItem*Role]] [] []
[2020-01-02 14:52:45] EXECENGINE.INFO: +++} [] []
[2020-01-02 14:52:45] EXECENGINE.INFO: ++} ExecEngine fixed 5 violations for rule 'Add navItemRoles for parent items' [] []
[2020-01-02 14:52:45] EXECENGINE.INFO: {++ ExecEngine fixing 9 violations for rule 'Fix TOT for isPartOf[PF_NavMenuItem*PF_NavMenu]' [] []
[2020-01-02 14:52:45] EXECENGINE.INFO: {+++ Fixing violation 1/9: ((Accounts[PF_NavMenuItem],MainMenu[PF_NavMenu])) [] []
[2020-01-02 14:52:45] EXECENGINE.INFO: InsPair(isPartOf,PF_NavMenuItem,Accounts,PF_NavMenu,MainMenu) [] []
[2020-01-02 14:52:45] EXECENGINE.DEBUG: Added (Accounts[PF_NavMenuItem],MainMenu[PF_NavMenu])[isPartOf[PF_NavMenuItem*PF_NavMenu]] [] []
[2020-01-02 14:52:45] EXECENGINE.INFO: +++} [] []
[2020-01-02 14:52:45] EXECENGINE.INFO: {+++ Fixing violation 2/9: ((Edit_32_navigation_32_menu[PF_NavMenuItem],MainMenu[PF_NavMenu])) [] []
[2020-01-02 14:52:45] EXECENGINE.INFO: InsPair(isPartOf,PF_NavMenuItem,Edit_32_navigation_32_menu,PF_NavMenu,MainMenu) [] []
[2020-01-02 14:52:45] EXECENGINE.DEBUG: Added (Edit_32_navigation_32_menu[PF_NavMenuItem],MainMenu[PF_NavMenu])[isPartOf[PF_NavMenuItem*PF_NavMenu]] [] []
[2020-01-02 14:52:45] EXECENGINE.INFO: +++} [] []
[2020-01-02 14:52:45] EXECENGINE.INFO: {+++ Fixing violation 3/9: ((List_32_all_32_interfaces[PF_NavMenuItem],MainMenu[PF_NavMenu])) [] []
[2020-01-02 14:52:45] EXECENGINE.INFO: InsPair(isPartOf,PF_NavMenuItem,List_32_all_32_interfaces,PF_NavMenu,MainMenu) [] []
[2020-01-02 14:52:45] EXECENGINE.DEBUG: Added (List_32_all_32_interfaces[PF_NavMenuItem],MainMenu[PF_NavMenu])[isPartOf[PF_NavMenuItem*PF_NavMenu]] [] []
[2020-01-02 14:52:45] EXECENGINE.INFO: +++} [] []
[2020-01-02 14:52:45] EXECENGINE.INFO: {+++ Fixing violation 4/9: ((Login[PF_NavMenuItem],MainMenu[PF_NavMenu])) [] []
[2020-01-02 14:52:45] EXECENGINE.INFO: InsPair(isPartOf,PF_NavMenuItem,Login,PF_NavMenu,MainMenu) [] []
[2020-01-02 14:52:45] EXECENGINE.DEBUG: Added (Login[PF_NavMenuItem],MainMenu[PF_NavMenu])[isPartOf[PF_NavMenuItem*PF_NavMenu]] [] []
[2020-01-02 14:52:45] EXECENGINE.INFO: +++} [] []
[2020-01-02 14:52:45] EXECENGINE.INFO: {+++ Fixing violation 5/9: ((Logout[PF_NavMenuItem],MainMenu[PF_NavMenu])) [] []
[2020-01-02 14:52:45] EXECENGINE.INFO: InsPair(isPartOf,PF_NavMenuItem,Logout,PF_NavMenu,MainMenu) [] []
[2020-01-02 14:52:45] EXECENGINE.DEBUG: Added (Logout[PF_NavMenuItem],MainMenu[PF_NavMenu])[isPartOf[PF_NavMenuItem*PF_NavMenu]] [] []
[2020-01-02 14:52:45] EXECENGINE.INFO: +++} [] []
[2020-01-02 14:52:45] EXECENGINE.INFO: {+++ Fixing violation 6/9: ((MyAccount[PF_NavMenuItem],MainMenu[PF_NavMenu])) [] []
[2020-01-02 14:52:45] EXECENGINE.INFO: InsPair(isPartOf,PF_NavMenuItem,MyAccount,PF_NavMenu,MainMenu) [] []
[2020-01-02 14:52:45] EXECENGINE.DEBUG: Added (MyAccount[PF_NavMenuItem],MainMenu[PF_NavMenu])[isPartOf[PF_NavMenuItem*PF_NavMenu]] [] []
[2020-01-02 14:52:45] EXECENGINE.INFO: +++} [] []
[2020-01-02 14:52:45] EXECENGINE.INFO: {+++ Fixing violation 7/9: ((MyScripts[PF_NavMenuItem],MainMenu[PF_NavMenu])) [] []
[2020-01-02 14:52:45] EXECENGINE.INFO: InsPair(isPartOf,PF_NavMenuItem,MyScripts,PF_NavMenu,MainMenu) [] []
[2020-01-02 14:52:45] EXECENGINE.DEBUG: Added (MyScripts[PF_NavMenuItem],MainMenu[PF_NavMenu])[isPartOf[PF_NavMenuItem*PF_NavMenu]] [] []
[2020-01-02 14:52:45] EXECENGINE.INFO: +++} [] []
[2020-01-02 14:52:45] EXECENGINE.INFO: {+++ Fixing violation 8/9: ((Roles[PF_NavMenuItem],MainMenu[PF_NavMenu])) [] []
[2020-01-02 14:52:45] EXECENGINE.INFO: InsPair(isPartOf,PF_NavMenuItem,Roles,PF_NavMenu,MainMenu) [] []
[2020-01-02 14:52:45] EXECENGINE.DEBUG: Added (Roles[PF_NavMenuItem],MainMenu[PF_NavMenu])[isPartOf[PF_NavMenuItem*PF_NavMenu]] [] []
[2020-01-02 14:52:45] EXECENGINE.INFO: +++} [] []
[2020-01-02 14:52:45] EXECENGINE.INFO: {+++ Fixing violation 9/9: ((StudentScripts[PF_NavMenuItem],MainMenu[PF_NavMenu])) [] []
[2020-01-02 14:52:45] EXECENGINE.INFO: InsPair(isPartOf,PF_NavMenuItem,StudentScripts,PF_NavMenu,MainMenu) [] []
[2020-01-02 14:52:45] EXECENGINE.DEBUG: Added (StudentScripts[PF_NavMenuItem],MainMenu[PF_NavMenu])[isPartOf[PF_NavMenuItem*PF_NavMenu]] [] []
[2020-01-02 14:52:45] EXECENGINE.INFO: +++} [] []
[2020-01-02 14:52:45] EXECENGINE.INFO: ++} ExecEngine fixed 9 violations for rule 'Fix TOT for isPartOf[PF_NavMenuItem*PF_NavMenu]' [] []
[2020-01-02 14:52:45] EXECENGINE.INFO: +} Exec engine run finished [] []
[2020-01-02 14:52:45] EXECENGINE.INFO: {+ Run #{2} (auto rerun: true) [] []
[2020-01-02 14:52:45] EXECENGINE.DEBUG: Select exec engine '{ExecEngine}' [] []
[2020-01-02 14:52:45] EXECENGINE.INFO: +} Exec engine run finished [] []
[2020-01-02 14:52:45] EXECENGINE.INFO: ExecEngine finished [] []
[2020-01-02 14:52:45] TRANSACTION.INFO: Request to close transaction: 883871315 [] []
[2020-01-02 14:52:45] TRANSACTION.INFO: Checking invariant rules [] []
[2020-01-02 14:52:46] TRANSACTION.INFO: Rollback transaction 883871315, because invariant rules do not hold [] []
```

**Comment by hanjoosten** [comment]
I am surprised that there is no list of the violations in the log, inbetween these two lines:
```
TRANSACTION.INFO: Checking invariant rules [] []
TRANSACTION.INFO: Rollback transaction <transaction>, because invariant rules do not hold [] []
```
If I look at the corresponding code:

~~~.php
/**
     * Returns if invariant rules hold and notifies user of violations (if any)
     * Note! Only checks affected invariant rules
     *
     * @return bool
     */
    public function checkInvariantRules(): bool
    {
        $this->logger->info("Checking invariant rules");
        
        $affectedInvRules = array_filter($this->getAffectedRules(), function (Rule $rule) {
            return $rule->isInvariantRule();
        });

        $rulesHold = true;
        foreach (RuleEngine::getViolations($affectedInvRules) as $violation) {
            $rulesHold = false; // set to false if there is one or more violation
            $this->app->userLog()->invariant($violation); // notify user of broken invariant rules
        }

        return $rulesHold;
    }
~~~
I would say that in any case that the returned `$rulesHold` is false, there must be some logging. Very weird that there isn't any in Stef's log. 
@Michiel-s , Is there any way we could determine what rule(s) are violated?

**Comment by hanjoosten** [comment]
@stefjoosten , The next time you supply a log, please do such as text, not as a graphic. That enables the reader to search for specific items in the log.

**Comment by Michiel-s** [comment]
@stefjoosten you write that you used https://github.com/AmpersandTarski/RAP/commit/5010d66c5a3798249fecf8a886e90f7cf8852db0 but I can’t find this commit. Could you verify that you used the latest commit I pushed, because that one makes sure you use the correct prototype backend.
1834f01405e8c1fdb7b059b43ede04dfa8fbd2d6

**Comment by stefjoosten** [comment]
will do

**Comment by Michiel-s** [comment]
I ment this one from Rap: https://github.com/AmpersandTarski/RAP/commit/1834f01405e8c1fdb7b059b43ede04dfa8fbd2d6

**Comment by stefjoosten** [comment]
Aha, I missed that update. I did it now. It gives a similar result, but different logs. This is the error log immediately after starting:
```
sjo00577@BA92-C02T81JCGTDY RAP % tail -f !*
tail -f ~/volumes/rap3-logs/error-2020-01-02.log
[2020-01-02 16:50:04] RULEENGINE.DEBUG: Rule 'UNI arfPersonRef[AccRegForm*PersonRef]' holds [] []
[2020-01-02 16:50:04] APPLICATION.INFO: INVARIANT 'Edit_32_navigation_32_menu[PF_Interface],Edit_32_navigation_32_menu[PF_Interface]' RULE: 'TOT label[PF_Interface*PF_Label]' [] []
[2020-01-02 16:50:04] APPLICATION.INFO: INVARIANT 'List_32_all_32_interfaces[PF_Interface],List_32_all_32_interfaces[PF_Interface]' RULE: 'TOT label[PF_Interface*PF_Label]' [] []
[2020-01-02 16:50:04] TRANSACTION.INFO: Rollback transaction 49256732, because invariant rules do not hold [] []
[2020-01-02 16:50:04] DATABASE.INFO: Rollback mysql database transaction for Transaction 49256732 [] []
[2020-01-02 16:50:04] DATABASE.DEBUG: ROLLBACK [] []
[2020-01-02 16:50:04] DATABASE.INFO: 1117 queries executed in this transaction [] []
[2020-01-02 16:50:04] API.ERROR: Initial installation does not satisfy invariant rules. See log files [] []
[2020-01-02 16:50:06] PERFORMANCE.INFO: Peak memory used: 11.91 Mb [] []
[2020-01-02 16:50:06] PERFORMANCE.INFO: Execution time  : 4.55 Sec [] []
```

**Comment by stefjoosten** [comment]
By the way, the only change I have now is that the volumes in the docker-compose.yml are made visible on my laptopy, so I can copy/paste the text of log files in this issue.

**Comment by Michiel-s** [comment]
I think this is an easy one. To me it looks like the prototypecontext is not grinded with itself.
@hanjoosten can you verify that the prototypecontext is first added/merged with the userscript and after that grinded with the prototypecontext? If not, than this must be fixed.
I hope this is the last error related to the refactoring of the Ampersand compiler, because this stuff worked fine in the 3.x version.

**Comment by hanjoosten** [comment]
Hmm. Are you sure you mean this? I cannot imagine. If the last step is a grind, than you only have populations. That cannot be correct. Do you mean something like this:

1) Take the user's script and merge it with the script of prototypecontext.
2) Grind the result of 1) with prototypecontext
3) merge the result of 1) and 2) 

This would be implemented as

~~~.haskell
    enablePrototype :: BuildRecipe -> BuildRecipe
    enablePrototype x = three
      where prototypeContext = BuildRecipe (MetaScript PrototypeContext) []
            one = x `merge` prototypeContext
            two = x `merge` prototypeContext
                     `andThen` Grind PrototypeContext
            three = one `merge` two
~~~

**Comment by Michiel-s** [comment]
That’s exactly what I mean. You state it correct.

**Comment by hanjoosten** [comment]
Well, it didn't help :(

**Comment by Michiel-s** [comment]
Ok. But the fix was needed anyway. Thx👍

**Comment by Michiel-s** [comment]
I’ll look into this issue further tomorrow

**Comment by Michiel-s** [comment]
@hanjoosten and I investigated this issue thoroughly today and concluded that the bug is caused by the merge function of two context. Somehow population sets are treated as the same and removed when merging contexts.
@hanjoosten will look further to find a solution.

**Comment by hanjoosten** [comment]
@Michiel-s , I removed two Eq and Ord instances, and did some rewriting on the merge funcion. This helps at least for the original problem I had with missing pairs with the `name` relations. There still are other issues, as can be seen in [this log](https://travis-ci.org/AmpersandTarski/Ampersand/jobs/632453584#L4977). Nevertheless, I think it is a good idea to  bring this to development

**Comment by hanjoosten** [comment]
After some playing around with this issue, I think that the root cause of this trouble is the Eq and Ord on origin. This is wrong, and we should get rid if those instances. 
In the early days of Ampersand, this seemed correct. Every element in the P_structure had it's origin in a file. There were a few exceptions to this, so it didn't feel wrong to have Eq and Ord defined. 
However, currently we can have elements that come from meatgrinder stuff, and there are other sources of stuff. 
Having said this, I am going to remove both instances and let the Haskell compiler point out where that hurts. I expect that this will eventually solve this issue. If not, at least it will prevent issues in the future.

**Comment by hanjoosten** [comment]
@sjcjoosten and @stefjoosten, what is your opinion about this insight?

**Comment by stefjoosten** [comment]
#### Diagnosis
The missing menu items were caused by the erroneous execution of recipes (new code).
The erroneous execution of recipes was caused by a wrong use of fold.
This has been fixed in commit AmpersandTarski/Ampersand@a83c64552a8893da164f0e3c11e2cda3002cd6a6.

**Comment by stefjoosten** [comment]
#### Verification
I compiled RAP with the fixed version (Ampersand branch fixAtlasComplete) and found that the navbar was OK.
